src/layout/umap.c

```diff
@@ -91,28 +149,92 @@ static igraph_error_t igraph_i_umap_find_sigma(const igraph_vector_t *distances,
     return IGRAPH_SUCCESS;
 }
 
+/* Check the distances argument, which should be NULL or a vector of nonnegative numbers */
+static igraph_error_t igraph_i_umap_check_distances(const igraph_vector_t *distances, igraph_integer_t no_of_edges) {
+
+    if (distances == NULL) {
+        return IGRAPH_SUCCESS;
+    }
+
+    if (igraph_vector_size(distances) != no_of_edges) {
+        IGRAPH_ERROR("Distances must be the same number as the edges in the graph.", IGRAPH_EINVAL);
+    }
+
+    for (igraph_integer_t eid = 0; eid != no_of_edges; eid++) {
+        if (VECTOR(*distances)[eid] < 0) {
+            IGRAPH_ERROR("Distances cannot be negative.", IGRAPH_EINVAL);
+        } else if (isnan(VECTOR(*distances)[eid])) {
+            IGRAPH_ERROR("Distances cannot contain NaN values.", IGRAPH_EINVAL);
+        }
+    }
+
+    return IGRAPH_SUCCESS;
+}
 
-/* Convert the graph with distances into a probability graph with exponential decay */
-/* NOTE: this is funny because the distance was a correlation to start with... ?? */
-static igraph_error_t igraph_i_umap_find_prob_graph(const igraph_t *graph,
-        const igraph_vector_t *distances, igraph_vector_t *umap_weights) {
+
+/**
+ * \function igraph_layout_umap_compute_connectivities
```

Comment:
I read this doc section, but I cannot tell what "connectivity" means here, or what this function does at all. Can you explain concisely what is being calculated?

After that we should think about whether the word "connectivity" can be avoided. This word is abused too much and used for entirely unrelated things by different papers. I guess the UMAP paper uses this term?

Reply:
Given a graph with distances between vertices associated with the edges, UMAP wants to move those distances into a probability space. So it wants to know what's the probability that vertices a and b are *connected* (topologically) given there is an edge with distance d between them.

Now if a has only one neighbor and that is b, then either a will end up a singleton (which UMAP tries to avoid) or it's connected to b, i.e. *the connectivity associated with that edge (which has distance d) is 1.0". If a has other neighbors at shorter distance, then the probability that b is connected becomes lower, since a is already connected to something else with higher priority.

That's the idea in a nutshell. Connectivity is how they call it in the paper and in the original UMAP implementation, so I moved away from our previous lingo "umap weights" to "connectivity" to make it easier for our average user who comes from that perspective.

Reply:
Can you document how this transformation from "distance" to "connectivity" is done exactly?

Reply:
Unless it's too long, in which case we should probably just point the reader to the UMAP paper. I don't know whether the concept can be summarized in a few sentences.

Reply:
yeah, we can't explain all the nitty gritty details here. People should read the paper if they want to understand the ins and outs of UMAP, with formulae, examples, and all.

Reply:
I looked at the paper and searched for "connectivity". I was unable to find the concept you implement here.

Where is "connectivity" defined under this name in https://arxiv.org/abs/1802.03426, and how might one find out what it is?

Reply:
First you renormalize the distances via rho and sigma, see algorithms 2 and 3 in the paper. Then you compute the connectivity as:

![image](https://user-images.githubusercontent.com/1200640/199430659-8462beec-bd9b-4cce-84d8-ea0c9b2cc514.png)

Reply:
Note to self: that formula is on page 21 in the linked preprint.

Reply:
see e.g. https://github.com/lmcinnes/umap/blob/a7118a004b403b8cc9d9876854fd9d7f95f7ea94/umap/umap_.py#L151

Reply:
Yes, this is too complicated to describe in the docs, but there should still be enough information there to allow one to find the part of the paper which explains what this function does.

At this moment my biggest concern is the unusual usage of the term "connectivity". You use this to describe a quantity associated with edges, but the above paper does not use this term this way (am I missing something?). Did you borrow this term from another paper perhaps? Can this term be replaced with something that relates to the paper better?

Reply:
> if you actually want to go through the math

I don't have time to do that now, but I want to make sure that if I wanted to do it, or even just use this function, it would be sufficiently easy for me to determine what it does after reading the docs. It should at least have non-confusing naming and enough pointers to the paper.

Reply:
What name do you propose? I think connectivity is just fine but am open to concrete suggestions.

Reply:
I have to log off now. Does the paper not use any name for this quantity? Is this what it means by the "membership strength" of a 1-simplex?  Otherwise why can't we just use "weight"?

I am not aware of any comparable use the term "connectivity" anywhere, but maybe I'm missing something.

Reply:
I should note that it is a very common situation in network science that we have a quantity associated with edges so that a large value represents a weaker connection (commonly a length / idstance), but for some algorithm we need a quantity where a large value represents a stronger connection. In this case the transform to this second form can be called "weight" or to be more specific, "(connection) strength".

Do I understand it that we have a similar situation here?

We have a similar case in https://github.com/igraph/igraph/issues/2229 as well as in a community detection algorithm I'm implementing now (no PR yet). In these two cases I went with "length" and "weight".

Reply:
"umap_weight" is what we have in master ATM. I can easily revert to that. It's some kind of strength of the type you are mentioning, yes.

Reply:
I think that we are diverging from the original topic of the PR here, which would have been to fix an error in the sampling process. IMHO it is totally fine to merge this PR even if the terminology is not 100% correct as long as we eventually fix it in the future (and if it is part of the public API, we should of course merge it before a release is made).

Reply:
fair enough. What's happening is that @szhorvat had requested better docs for a while and I agreed to improve them a little. I think we are there now. I'll go back to the "umap_weight" terminology and then we can merge.

Reply:
This function _is_ part of the public API, and is added by this PR.

If it's merged as is, it'll likely stay as is, and any update will be forgotten or dismissed.

Reply:
btw I finally remembered where I borrowed the term "connectivity" from:

https://github.com/scverse/scanpy/blob/e5cdbbc02702d779352cbbdfa8e54020d93301ea/scanpy/tools/_umap.py#L107